File: kraken-engine/src/test/java/kraken/runtime/engine/handlers/AssertionPayloadHandlerTest.java
/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package kraken.runtime.engine.handlers;

import static kraken.runtime.engine.handlers.PayloadHandlerTestConstants.SESSION;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

import java.util.List;

import org.junit.Before;
import org.junit.Test;

import kraken.TestRuleBuilder;
import kraken.runtime.engine.context.data.DataContext;
import kraken.runtime.engine.result.AssertionPayloadResult;
import kraken.runtime.expressions.KrakenExpressionEvaluator;
import kraken.runtime.model.rule.RuntimeRule;

public class AssertionPayloadHandlerTest {
    private AssertionPayloadHandler assertionPayloadHandler;

    @Before
    public void setUp() {
        assertionPayloadHandler = new AssertionPayloadHandler(new KrakenExpressionEvaluator());
    }

    @Test
    public void executePayloadShouldReturnSuccessfulResultWhenExpressionIsTrue() {
        Coverage coverage = new Coverage();
        coverage.setLevel(7);
        DataContext dataContext = new DataContext();
        dataContext.setDataObject(coverage);

        final RuntimeRule rule = TestRuleBuilder.getInstance().assertionPayload("level > 5", List.of("Level: ", ""), List.of("level")).build();
        AssertionPayloadResult payloadResult = (AssertionPayloadResult) assertionPayloadHandler.executePayload(
                rule.getPayload(),
                rule,
                dataContext,
                SESSION
        );

        assertThat(payloadResult.getSuccess(), is(true));
        assertThat(payloadResult.getMessage(), is(equalTo("Level: 7")));
    }

    @Test
    public void executePayloadShouldReturnUnsuccessfulResultWhenExpressionIsFalse() {
        Coverage coverage = new Coverage();
        coverage.setLevel(3);
        DataContext dataContext = new DataContext();
        dataContext.setDataObject(coverage);
        final RuntimeRule rule = TestRuleBuilder.getInstance().assertionPayload("level > 5", List.of("Level: ", ""), List.of("level")).build();
        AssertionPayloadResult payloadResult = (AssertionPayloadResult) assertionPayloadHandler.executePayload(
                rule.getPayload(),
                rule,
                dataContext,
                SESSION
        );

        assertThat(payloadResult.getSuccess(), is(false));
        assertThat(payloadResult.getMessage(), is(equalTo("Level: 3")));
    }
}

File: javascript/packages/engine/__tests__/engine/hadlers/AssertionPayloadHandler.test.ts
/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { AssertionPayloadResult } from 'kraken-engine-api'
import { Payloads } from 'kraken-model'
import { PayloadBuilder, RulesBuilder } from 'kraken-model-builder'

import { DataContext } from '../../../src/engine/contexts/data/DataContext'
import { AssertionPayloadHandler } from '../../../src/engine/handlers/AssertionPayloadHandler'
import { payloadResultTypeChecker } from '../../../src/engine/results/PayloadResultTypeChecker'
import { mock } from '../../mock'

let dataContext: DataContext
const handler = new AssertionPayloadHandler(mock.evaluator)
const { session } = mock
const { Policy } = mock.modelTreeJson.contexts

beforeEach(() => {
    dataContext = mock.data.dataContextEmpty()
})
describe('assertionPayloadHandler', () => {
    it('should create instance', () => {
        expect(handler.handlesPayloadType()).toBe(Payloads.PayloadType.ASSERTION)
    })
    it('should return payload payloadResult with type AssertionPayloadResult', () => {
        const rule = RulesBuilder.create()
            .setName('R001')
            .setContext(Policy.name)
            .setTargetPath(Policy.fields.state.name)
            .setPayload(PayloadBuilder.asserts().that('false'))
            .build()
        const result = handler.executePayload(rule, dataContext, session)
        expect(payloadResultTypeChecker.isAssertion(result)).toBeTruthy()
    })
    it('should execute payload', () => {
        const rule = RulesBuilder.create()
            .setName('R001')
            .setContext(Policy.name)
            .setTargetPath(Policy.fields.state.name)
            .setPayload(PayloadBuilder.asserts().that('__dataObject__.state != null', 'assertion failed'))
            .build()
        const result = handler.executePayload(rule, dataContext, session) as AssertionPayloadResult
        expect(result.message?.errorMessage).toBe('assertion failed')
        expect(result.success).toBeFalsy()
    })
    it('should execute payload with no err message and return false', () => {
        const ruleNoError = RulesBuilder.create()
            .setName('R001')
            .setContext(Policy.name)
            .setTargetPath(Policy.fields.state.name)
            .setPayload(PayloadBuilder.asserts().that('false', 'message'))
            .build()
        const result = handler.executePayload(ruleNoError, dataContext, session) as AssertionPayloadResult
        expect(result.message?.errorMessage).toBe('message')
        expect(result.success).toBeFalsy()
    })
})
